metagraph/src/annotation/binary_matrix/row_diff/row_diff.hpp

```diff
@@ -9,92 +9,175 @@
+ *
+ * RowDiff sparsification can be applied to any BinaryMatrix instance.
  * The row-diff binary matrix is defined by three data structures:
- *   1. #diffs_ stores the concatenated diffs row by row. Since each row-diff has variable
- *      length, the boundary_ vector marks the end of each row-diff.
- *   2. boundary_ marks the end of a row diff in #diffs_
- *   3. terminal_ Rows marked as terminal are stored in full.
- * Retrieving data from RowDiff requires the associated #graph_. In order to get the i-th
- * row, we start traversing the node corresponding to i in #graph_ and accumulate the
- * diffs until we hit a terminal node, which is stored in full.
+ *   1. #diffs_ the underlying sparsified (diffed) #BinaryMatrix
+ *   2. #terminal_ rows marked as terminal are stored in full
+ *   3. #graph_ the graph that was used to determine adjacent rows for sparsification
+ * Retrieving data from RowDiff requires the associated #graph_. In order to get the
+ * annotation for  i-th row, we start traversing the node corresponding to i in #graph_
+ * and accumulate the values in #diffs until we hit a terminal node, which is stored in
+ * full.
  */
+//NOTE: Clang aggressively abuses the clause in the C++ standard (14.7.1/11) that allows
+// virtual methods in template classes to not be instantiated if unused and mistakenly
+// does not instantiate the virtual methods in this class, so I had to move definitions
+// to the header (gcc works fine)
+ template <class BaseMatrix>
 class RowDiff : public BinaryMatrix {
   public:
     RowDiff() {}
 
-    RowDiff(const uint64_t num_columns,
-            const uint64_t num_relations,
-            const graph::DBGSuccinct *graph,
-            const sdsl::enc_vector<> &diffs,
-            const sdsl::bit_vector &boundary,
-            const sdsl::bit_vector &terminal)
-            : num_columns_(num_columns),
-              num_relations_(num_relations),
-              graph_(graph),
-              diffs_(diffs),
-              boundary_(boundary),
-              terminal_(terminal) {
-        sdsl::util::init_support(sboundary_, &boundary_);
+    RowDiff(const graph::DBGSuccinct *graph,
+               BaseMatrix &&diffs,
+               const std::string &terminal_file)
+        : graph_(graph), diffs_(std::move(diffs)), terminal_file_(terminal_file) {
+        load_terminal(terminal_file_, &terminal_);
     }
 
-    uint64_t num_columns() const override { return num_columns_; }
+    uint64_t num_columns() const override { return diffs_.num_columns(); }
+
+    const std::string& terminal_file() const { return terminal_file_; }
+
+    const graph::DBGSuccinct *graph() const { return graph_; }
 
     /**
      * Returns the number of set bits in the matrix.
      */
-    uint64_t num_relations() const override { return num_relations_; }
-
-    uint64_t num_rows() const override { return terminal_.size(); }
+    uint64_t num_relations() const override { return diffs_.num_relations(); }
```

Comment:
move the implementation out of the class to the bottom of the file.
e.g.
```c++
template <class BaseMatrix>
inline bool RowDiff<BaseMatrix>::get(Row row, Column column) const {
```

Reply:
It's good to be able to see the pure interface. Currently, it's crowded with the implementations which make it hard to view the list of all methods in the class. This won't change the performance, just make .hpp more structured.